### PR TITLE
Workaround for docker cp ownership settings 

### PR DESCRIPTION
**Pull request by lvotypko**
I am opening this pull request to have a discussion about docker cp functionality. It does not work well in case docker daemon. The situation:
whole ATH is running inside container - so container A is created for it. ATH needs for its execution another container so it creates container B by docker daemon.
Now the container A calls docker cp by docker daemon to copy file from container B. In this case the file from container B is copied into container A, but the ownership of the file is assigned to root. 

It is impossible to do something with this file if rights are not open for another user. For example this test is failing because of it https://ci.jenkins.io/job/Core/job/acceptance-test-harness/job/master/64/testReport/plugins/FtpPublishPluginTest/publish_resources_and_remove_prefix/
I know that my solution is not something nice.... it is pretty dirty (to use sudo chown under non admin user), but it is the only one solution which I found that does not decrease whole funkcionality of the method and does not require something complicated (it enables copying of directories and all kind of files). We can use for example cat command or another way how to copy the content of file, but the behavior of method will need to be changed (to copy only files not directories) and I do not think it will work for any kind of file. 

It does not seem that there is an intention to change this behavior soon.
https://forums.docker.com/t/can-docker-cp-set-user-id-and-group/21562

Any ideas how to deal with it?

### CODE REVIEW

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 1]
```diff
@@ -191,6 +194,17 @@ public boolean cp(String from, String toPath) {
...
             Docker.cmd("cp").add(cid + ":" + from).add(new File(toPath))
                     .popen().verifyOrDieWith(format("Failed to copy %s to %s", from, toPath));
+            if(sharingHostDockerService()){
+                //docker deaomon causes that cp set owner root, try to change owner
+                String user = System.getProperty("user.name");
+                String owner = Files.getOwner(Paths.get(destFile.getAbsolutePath())).getName();
+                if(!user.equals(owner)) {
+                    String id = InetAddress.getLocalHost().getHostName();
+                    Docker.cmd("exec").add(id,"sudo", "chown", "-R", user, destFile.getAbsoluteFile())
+                            .popen().verifyOrDieWith(format("Failed to copy %s to %s", from, toPath));
```
Comment (olivergondza) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: "Failed to chown..."

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 2]
```diff
@@ -191,6 +194,17 @@ public boolean cp(String from, String toPath) {
         try {
             Docker.cmd("cp").add(cid + ":" + from).add(new File(toPath))
                     .popen().verifyOrDieWith(format("Failed to copy %s to %s", from, toPath));
+            if(sharingHostDockerService()){
+                //docker deaomon causes that cp set owner root, try to change owner
```
Comment (olivergondza) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: typo: daemon

**Comment by lvotypko** [comment]
@jglick @olivergondza I opened this mostly for evaluation of problem. Do you think it is good idea to do it this way - I mean use sudo chown. It is dirty solution and there is still option to remove cp method, or declare it as deprecated - not a nice one as well because this method is there for a long time and is used. On the other hand when you look in ATH tests, it is not use a lot and it can be easily replaced, viz https://github.com/jenkinsci/acceptance-test-harness/pull/436. Since this pull request was already merged, we have ideal time to remove/deprecate it. When we use my workaround, we can have future we can let this functionality there but we can have troubles with breaking the security by giving sudo to chown operation. I have no idea how it can be an issue now (for the purpose we use it), but who knows how it will be used several years later. Now I would prefer to declare it as deprecated and remove all usage of this method form ATH tests. Do you agree? Or it seems better this fix (code I posted, of course with required changes by Oliver)? What is your opinion? Thanks!

**Comment by olivergondza** [comment]
Hmm, since we have found a way not to need this, I am ok to either keep things the way they are or deprecate it. Adding a fix that complicates things this way without having a client that would depend on such fix is something I would like to avoid.

**Comment by lvotypko** [comment]
@jglick Can you merge it please?

**Pull request merged**